Fixed derivative handling of pid

saarnold opened the pull request:
In cases where the input value stays constant, and the reference value is changing, the previous implementation would lead to the derivative value not being used.

To calculate the derivative, only the input value was used, instead of the full error value (e.g. the reference - input).

Changed the implementation to store the error instead of the previous input value, and use that to calculate the derivative.

I'm citing @jakobs here ;)

jakobs committed:
add getters for modes

saarnold committed:
removed dead code

saarnold committed:
fixed derivate handling of PID

In cases where the input value stays constant, and the reference value
is changing, the previous implementation would lead to the derivative
value not being used.

To calculate the derivative, only the input value was used, instead of
the full error value (e.g. the reference - input).

Changed the implementation to store the error instead of the previous
input value, and use that to calculate the derivative.

ajishbabu commented:
hi Sacha, the reference input is removed intentionally from the calculation to avoid jumps in case of non-smooth reference inputs (step input). Just using the derivative of measured value should normally serve the purpose.

doudou commented:
@ajishbabu: how does that work in case where you have a varying command ? Let's say that you have a x=y command and a x=y+1 system response, the derivative term would kick in (since d(response) != 0) even though it's not supposed to (d(error) == 0)

ajishbabu commented:
@doudou the derivative action in this implementation just tries to slow the system down if it is going too fast. It does not directly react to the input command.

rafaelsaback commented:
Hey guys, this approach of applying the derivative action only to the measurement is quite common, especially when it comes to "constant" references. In this "constant" references, the setpoint doesn't change in a continuous way, usually it's kept stable in one value for some period of time, and then it's changed suddenly (e.g. from 1 to 2). In this case, the derivative action would have an infinite value, and this is not good for the control.

On the other hand, if we have a continuously varying reference (maybe it's the waypoint following case, I don't know how it works exactly), then applying the derivative to the error wouldn't be any problem.

So, in my point of view, the choice of applying the derivative action to the error or only to the measurement  should be made by the user, who is the one that (theoretically) knows best the particular system.

doudou commented:
+1 to @rafaelsaback remark. The behaviour should definitely be chosen at configuration time.

jakobs commented:
The case where this popped up, was a controller, which had a constant measurement and a varying setpoint. This was to control the vehicle's heading in the yaw fixed frame. In this frame, the heading is always 0. We where wondering why  the D part of the controller never kicked in. 

A assumed this was an error in the implementation when I found the position in the code. Wikipedia on PID controllers says that the derivative of the error is used. If this is important in some cases to only use the derivative of the measurement, I would also say this should be a parameter. But it should be off by default.

Commit:
Now the derivative can be applied either to the error or only to the output.

rafaelsaback commented:
I've made the changes we have discussed previously, and basically, now the user can choose between aplying the derivative action to the error or only to the output.

As default, I followed Jakob suggestion and the derivative action is applied to the error.

doudou commented:
I would rather keep the flag to off (== old behaviour). Otherwise, tuned controllers will all of a sudden start to behave differently - and the reason will be hard to spot to say the least.

Commit:
Changed the default derivative behavior

The default derivative behavior was changed to the old fixed behavior, in which the derivative is applied only to the output.

rafaelsaback commented:
Since I don't have the rights to push directly to the repo, I've created a pull request to this branch (check the other pull request).

The default behavior was changed to the old one (derivative of the output).

doudou commented:
WIth @rafaelsaback latest PR, this would look good to me ... @ajishbabu ?

ajishbabu committed:
Merge pull request #5 from rafaelsaback/patch-1

Changed the default derivative behavior

doudou commented:
@ajishbabu: usually, don't hesitate deleting the PR branch if github proposes it to you (just after the merge you can have a "Delete branch" button)

ajishbabu commented:
sure ... next time
